src/lib/operators/partitioning.cpp

```diff
@@ -0,0 +1,301 @@
+    auto& values = casted_target->values();
+
+    auto target_is_nullable = casted_target->is_nullable();
+
+    if (auto casted_source = std::dynamic_pointer_cast<const ValueColumn<T>>(source)) {
+      size_t target_index = target_start_index;
+      for (size_t row_to_copy : rows_to_copy) {
+        if (row_to_copy >= source_start_index) {
+          std::copy_n(casted_source->values().begin() + row_to_copy, 1, values.begin() + target_index);
+          target_index++;
+        }
+      }
+
+      if (casted_source->is_nullable()) {
+        // Values to insert contain null, copy them
+        if (target_is_nullable) {
+          target_index = target_start_index;
+          for (size_t row_to_copy : rows_to_copy) {
+            if (row_to_copy >= source_start_index) {
+              std::copy_n(casted_source->null_values().begin() + row_to_copy, 1,
+                          casted_target->null_values().begin() + target_index);
+              target_index++;
+            }
+          }
+        } else {
+          for (const auto null_value : casted_source->null_values()) {
+            Assert(!null_value, "Trying to insert NULL into non-NULL column");
+          }
+        }
+      }
+    } else if (auto casted_dummy_source = std::dynamic_pointer_cast<const ValueColumn<int32_t>>(source)) {
+      // We use the column type of the Dummy table used to insert a single null value.
+      // A few asserts are needed to guarantee correct behaviour.
+      Assert(rows_to_copy.size() == 1, "Cannot insert multiple unknown null values at once.");
+      Assert(casted_dummy_source->size() == 1, "Source column is of wrong type.");
+      Assert(casted_dummy_source->null_values().front() == true, "Only value in dummy table must be NULL!");
+      Assert(target_is_nullable, "Cannot insert NULL into NOT NULL target.");
+
+      // Ignore source value and only set null to true
+      casted_target->null_values()[target_start_index] = true;
+    } else {
+      // } else if(auto casted_source = std::dynamic_pointer_cast<ReferenceColumn>(source)){
+      // since we have no guarantee that a referenceColumn references only a single other column,
+      // this would require us to find out the referenced column's type for each single row.
+      // instead, we just use the slow path below.
+
+      size_t target_index = target_start_index;
+      for (size_t row_to_copy : rows_to_copy) {
+        if (row_to_copy >= source_start_index) {
+          auto ref_value = (*source)[row_to_copy];
+          if (variant_is_null(ref_value)) {
+            Assert(target_is_nullable, "Cannot insert NULL into NOT NULL target");
+            values[target_index] = T{};
+            casted_target->null_values()[target_index] = true;
+          } else {
+            values[target_index] = type_cast<T>(ref_value);
+          }
+          target_index++;
+        }
+      }
+    }
+  }
+};
+
+Partitioning::Partitioning(const std::string& table_to_partition_name,
```

Comment:
Is this really an operator? It has neither input nor output and cannot be part of a query plan. What do you gain from it being an operator instead of a free class?